core/src/storage/flush.rs

```diff
@@ -31,7 +31,83 @@
 /// All abstractions built upon them that do not have their own caching mechanism
 /// shall simply forward flushing to their interiors. Examples for this are
 /// `storage::Vec` or `storage::Value`.
+
+use parity_codec::Encode;
+use crate::storage::key::Key;
+
 pub trait Flush {
     /// Flushes the cached state back to the contract storage, if any.
-    fn flush(&mut self);
+    fn flush(&mut self) {
+        unreachable!();
+    }
+
+    /// Default implementation which forwards to flush.
+    /// This realizes recursive behavior for e.g. nested vectors.
+    fn flush_at(&mut self, _at: Key) {
+        self.flush();
+    }
+}
+
+impl Flush for i8 where Self: Encode {
```

Comment:
Why do we need the `Self: Encode` btw?

Reply:
Also I think you can get rid of lots of code dupe with a decent `macro_rules!` macro. ;)